### PR TITLE
Migrate to PHP 8: AbstractQuery and child classes

### PR DESCRIPTION
**Pull request by greg0ire**
Blocked by 
- https://github.com/doctrine/orm/pull/9777
- https://github.com/doctrine/orm/pull/9809

TODO:
- [x] merge the above
- [x] merge up the patch branch
- [x] merge up the minor branch
- [x] run rector

### CODE REVIEW

File: lib/Doctrine/ORM/AbstractQuery.php [hunk 1]
```diff
@@ -214,59 +193,52 @@ public function setCacheRegion($cacheRegion)
...
      * Sets the life-time for this query into second level cache.
      *
-     * @param int $lifetime
-     *
      * @return $this
      */
-    public function setLifetime($lifetime)
+    public function setLifetime(int $lifetime): static
     {
         $this->lifetime = (int) $lifetime;
```
Comment (derrabus) [inline lib/Doctrine/ORM/AbstractQuery.php]:
```suggestion
        $this->lifetime = $lifetime;
```

File: lib/Doctrine/ORM/AbstractQuery.php [hunk 2]
```diff
@@ -310,7 +278,7 @@ public function free()
      * @return ArrayCollection The defined query parameters.
```
Comment (derrabus) [inline lib/Doctrine/ORM/AbstractQuery.php]:
```suggestion
```

File: lib/Doctrine/ORM/AbstractQuery.php [hunk 3]
```diff
@@ -322,7 +290,7 @@ public function getParameters()
      *
      * @return Parameter|null The value of the bound parameter, or NULL if not available.
      */
-    public function getParameter($key)
+    public function getParameter($key): ?Parameter
```
Comment (derrabus) [inline lib/Doctrine/ORM/AbstractQuery.php]:
```suggestion
    public function getParameter(mixed $key): ?Parameter
```

At least that's what the doc block says. Can we narrow that down to `int|string`?

Comment (greg0ire) [reply lib/Doctrine/ORM/AbstractQuery.php]: That would make sense 👍

Comment (greg0ire) [reply lib/Doctrine/ORM/AbstractQuery.php]: Incorporated that in #9777

File: lib/Doctrine/ORM/AbstractQuery.php [hunk 4]
```diff
@@ -217,59 +196,52 @@ public function setCacheRegion($cacheRegion)
...
      * @return string|null The cache region name; NULL indicates the default region.
      */
-    public function getCacheRegion()
+    public function getCacheRegion(): ?string
     {
         return $this->cacheRegion;
     }
 
     /**
      * @return bool TRUE if the query cache and second level cache are enabled, FALSE otherwise.
```
Comment (derrabus) [inline lib/Doctrine/ORM/AbstractQuery.php:200]:
```suggestion
     * Checks if the query cache and second level cache are enabled.
```

File: lib/Doctrine/ORM/AbstractQuery.php [hunk 5]
```diff
@@ -217,59 +196,52 @@ public function setCacheRegion($cacheRegion)
      *
      * @return string|null The cache region name; NULL indicates the default region.
```
Comment (derrabus) [inline lib/Doctrine/ORM/AbstractQuery.php:192]:
```suggestion
     * Returns the cache region name or NULL to indicate the default region.
```

File: lib/Doctrine/ORM/AbstractQuery.php (hunk 3 above)
Comment (derrabus) [reply lib/Doctrine/ORM/AbstractQuery.php]:
```suggestion
    public function getParameter(int|string $key): ?Parameter
```

File: lib/Doctrine/ORM/AbstractQuery.php [hunk 6]
```diff
@@ -348,7 +315,7 @@ static function (Query\Parameter $parameter) use ($key): bool {
      *
      * @return $this
      */
-    public function setParameters($parameters)
+    public function setParameters(ArrayCollection|array $parameters): static
```
Comment (derrabus) [inline lib/Doctrine/ORM/AbstractQuery.php:309]: We should try to narrow this down in 2.13.

File: lib/Doctrine/ORM/AbstractQuery.php [hunk 7]
```diff
@@ -395,14 +362,12 @@ public function setParameter($key, $value, $type = null)
     /**
      * Processes an individual parameter value.
      *
-     * @param mixed $value
-     *
      * @return mixed[]|string|int|float|bool|object|null
      * @psalm-return array|scalar|object|null
```
Comment (derrabus) [inline lib/Doctrine/ORM/AbstractQuery.php]: Do these annotations provide any real value? They're hardly more specifiy than `mixed`.

File: lib/Doctrine/ORM/AbstractQuery.php [hunk 8]
```diff
@@ -576,7 +532,7 @@ public function getHydrationCacheProfile()
      *
      * @return $this
      */
-    public function setResultCacheProfile(?QueryCacheProfile $profile = null)
+    public function setResultCacheProfile(?QueryCacheProfile $profile = null): static
```
Comment (derrabus) [inline lib/Doctrine/ORM/AbstractQuery.php]: Can we remove the default here? That looks like a PHP5 hack to me.

File: lib/Doctrine/ORM/AbstractQuery.php [hunk 9]
```diff
@@ -598,10 +554,8 @@ public function setResultCacheProfile(?QueryCacheProfile $profile = null)
 
     /**
      * Defines a cache driver to be used for caching result sets and implicitly enables caching.
-     *
-     * @return $this
      */
-    public function setResultCache(?CacheItemPoolInterface $resultCache = null)
+    public function setResultCache(?CacheItemPoolInterface $resultCache = null): static
```
Comment (derrabus) [inline lib/Doctrine/ORM/AbstractQuery.php]: Same here.

File: lib/Doctrine/ORM/AbstractQuery.php [hunk 10]
```diff
@@ -949,14 +870,12 @@ public function getHints()
      * Executes the query and returns an iterable that can be used to incrementally
      * iterate over the result.
      *
-     * @param ArrayCollection|array|mixed[] $parameters    The query parameters.
-     * @param string|int|null               $hydrationMode The hydration mode to use.
      * @psalm-param ArrayCollection<int, Parameter>|mixed[] $parameters
```
Comment (derrabus) [inline lib/Doctrine/ORM/AbstractQuery.php:849]: It's a bit odd that we use `iterable` as native type, but a more specific `ArrayCollection` as annotated type. I think, we should fix the annotation.

**Review by derrabus — CHANGES_REQUESTED** [review]

**Comment by greg0ire** [comment]
@derrabus I'm AFK until Sunday, feel free to push to my branch, I agree with all of your comments.

File: lib/Doctrine/ORM/AbstractQuery.php (hunk 8 above)
Comment (derrabus) [reply lib/Doctrine/ORM/AbstractQuery.php]: I'm paving the way with #9791.

Comment (derrabus) [reply lib/Doctrine/ORM/AbstractQuery.php]: #9798

File: lib/Doctrine/ORM/AbstractQuery.php (hunk 6 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/AbstractQuery.php:309]: :thinking: what do you mean? Drop the `ArrayCollection|` part? I know I have never used it that way.

File: lib/Doctrine/ORM/AbstractQuery.php (hunk 7 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/AbstractQuery.php]: I'll try widening to `mixed` on 2.13.x, see if that impacts static analysis.

File: lib/Doctrine/ORM/AbstractQuery.php (hunk 10 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/AbstractQuery.php:849]: I agree it's a problem, but I don't think fixing the annotation is the right thing to do: down the line, we call `setParameters`, which can deal with `mixed[]`, but not with `ArrayCollection<stdClass>`. It needs to be `ArrayCollection<Parameter>`.

**Review by SenseException — DISMISSED** [review]

File: lib/Doctrine/ORM/AbstractQuery.php (hunk 10 above)
Comment (derrabus) [reply lib/Doctrine/ORM/AbstractQuery.php:849]: You're right. But then, we should have the same native parameter type as `setParameters()`, so `iterable` should be `ArrayCollection|array`.

File: lib/Doctrine/ORM/Query.php [hunk 11]
```diff
@@ -479,16 +451,12 @@ public function useQueryCache($bool): self
     /**
      * Defines how long the query cache will be active before expire.
      *
-     * @param int $timeToLive How long the cache entry is valid.
+     * @param int|null $timeToLive How long the cache entry is valid.
```
Comment (derrabus) [inline lib/Doctrine/ORM/Query.php]: Should be fixed on 2.13.x as well.

File: lib/Doctrine/ORM/Query.php [hunk 12]
```diff
@@ -509,7 +477,7 @@ public function getQueryCacheLifetime(): ?int
      *
      * @return $this
      */
-    public function expireQueryCache($expire = true): self
+    public function expireQueryCache(bool $expire = true): self
```
Comment (derrabus) [inline lib/Doctrine/ORM/Query.php:479]: We can delete the corresponding `@param`, I think.

File: lib/Doctrine/ORM/Query.php [hunk 13]
```diff
@@ -537,7 +505,7 @@ public function free(): void
      *
      * @param string|null $dqlQuery DQL Query.
```
Comment (derrabus) [inline lib/Doctrine/ORM/Query.php]:
```suggestion
```

File: lib/Doctrine/ORM/Query.php [hunk 14]
```diff
@@ -597,7 +565,7 @@ public function contains($dql): bool
      *
      * @return $this
      */
-    public function setFirstResult($firstResult): self
+    public function setFirstResult(?int $firstResult): self
```
Comment (derrabus) [inline lib/Doctrine/ORM/Query.php:565]: Shall we disallow `null`, to be consistent with the DBAL?

File: lib/Doctrine/ORM/AbstractQuery.php (hunk 10 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/AbstractQuery.php:849]: Yes, definitely, I can change that in this PR 👍

File: lib/Doctrine/ORM/AbstractQuery.php (hunk 7 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/AbstractQuery.php]: https://github.com/doctrine/orm/pull/9801

File: lib/Doctrine/ORM/Query.php (hunk 11 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/Query.php]: "fixed"? 2.12.x then.

File: lib/Doctrine/ORM/Query.php (hunk 14 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/Query.php:565]: Yes, that would make sense. That will need to be initiated on 2.13.x

File: lib/Doctrine/ORM/AbstractQuery.php [hunk 15]
```diff
@@ -712,24 +654,10 @@ public function getQueryCacheProfile()
...
-        if (! in_array($fetchMode, [Mapping\ClassMetadata::FETCH_EAGER, Mapping\ClassMetadata::FETCH_LAZY], true)) {
-            Deprecation::trigger(
-                'doctrine/orm',
-                'https://github.com/doctrine/orm/pull/9777',
-                'Calling %s() with something else than ClassMetadata::FETCH_EAGER or ClassMetadata::FETCH_LAZY is deprecated.',
-                __METHOD__
-            );
-            $fetchMode = Mapping\ClassMetadata::FETCH_LAZY;
-        }
-
```
Comment (greg0ire) [inline lib/Doctrine/ORM/AbstractQuery.php:731]: From https://github.com/doctrine/orm/pull/9777#discussion_r879878197 : 

> On 3.0, we would throw if the caller passed anything but the two allowed values, right?

Do we really want to do this? If yes, maybe we should require an assertion library :thinking:

File: lib/Doctrine/ORM/AbstractQuery.php [hunk 16]
```diff
@@ -86,93 +83,74 @@ abstract class AbstractQuery
...
      * Initializes a new instance of a class derived from <tt>AbstractQuery</tt>.
      */
-    public function __construct(EntityManagerInterface $em)
-    {
-        $this->_em        = $em;
+    public function __construct(
+        /**
+         * The entity manager used by this query object.
+         */
+        protected EntityManagerInterface $_em
```
Comment (derrabus) [inline lib/Doctrine/ORM/AbstractQuery.php]: I'd rather rename the property to `em` than adding that `_` prefix to the constructor parameter, tbh.

File: lib/Doctrine/ORM/AbstractQuery.php [hunk 17]
```diff
@@ -325,12 +287,12 @@ public function getParameters()
...
      */
-    public function getParameter($key)
+    public function getParameter(int|string $key): ?Parameter
     {
-        $key = Query\Parameter::normalizeName($key);
+        $key = Parameter::normalizeName($key);
 
         $filteredParameters = $this->parameters->filter(
-            static function (Query\Parameter $parameter) use ($key): bool {
+            static function (Parameter $parameter) use ($key): bool {
```
Comment (derrabus) [inline lib/Doctrine/ORM/AbstractQuery.php]:
```php
$filteredParameters = $this->parameters->filter(
    static fn (Parameter $parameter): bool => $parameter->getName() === $key
);
```

File: lib/Doctrine/ORM/AbstractQuery.php (hunk 16 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/AbstractQuery.php]: Me too, that was me not questioning Rector's work.

**Review by derrabus — DISMISSED** [review]

**Pull request merged**